src/app/decorations/qgsdecorationnortharrow.cpp

```diff
@@ -168,8 +168,14 @@ void QgsDecorationNorthArrow::render( const QgsMapSettings &mapSettings, QgsRend
                                      ( centerYDouble * std::cos( radiansDouble ) )
                                    ) - centerYDouble );
     // need width/height of paint device
-    int deviceHeight = context.painter()->device()->height();
```

Comment:
Is it worth grepping for other use of device()->height()? I think other parts of code use this (inverted polygon renderer comes to mind)

Reply:
it feels strange that the device has the size of the image, no?
Shouldn't it be scaled by the DPR of the image?

A quick run of grep gave me the heat map renderer, but not inverted polygon renderer.